subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java

```diff
@@ -97,12 +99,22 @@ private void addJacocoConfigurations() {
         agentConf.setVisible(false);
         agentConf.setTransitive(true);
         agentConf.setDescription("The Jacoco agent to use to get coverage data.");
+        agentConf.attributes(attributes -> attributes.attribute(ArtifactAttributes.ARTIFACT_FORMAT, "jacoco-agent-jar"));
```

Comment:
Regarding resolving: Yes and no. It works because you can have a configuration that is both resolvable and also everything else. Which is the default for backwards compatibility. So although the suggestion is to separate the configuration on which you declare dependencies and the one you resolve, this here works. And I think it is ok for this internal usage. If we ever forbid this, than we can adjust it (together with the trillion other places we probably do this).

You can also do this with an arifact view, but adding the attribute as _requested_ on the configuration also works. Since the configuration is used for nothing else, this is fine I think.

You are right that the difference is that `ARTIFACT_FORMAT= jacoco-agent-jar` will also be considered during variant matching (which is done before selecting/transforming artifacts). However, the attribute here is basically ignored during matching, because there is no conflict. We are in the "Java world" and in that world we have rules that always default to the "Java runtime" variant if there is no explicit selection of a "Usag" and no conflict between other attributes (of which non is requested here).

So... maybe a artifact view would be "cleaner". But from my perspective the current solution is fine, because we really use this as a tool to get hold of, and transform, a **single** artifact.